orchestra/todos/import_export.py

```diff
@@ -0,0 +1,122 @@
+    return 'https://docs.google.com/spreadsheets/d/{}'.format(sheet['id'])
+
+
+def export_to_spreadsheet(todo_list_template):
+    """
+    Recursively descend down the to-do tree of a `todo_list_template`
+    and add a row per to-do to a CSV, uploading it to a Google Sheet
+    when done. The first two columns of the sheet capture json-encoded
+    remove_if and skip_if logic. After that, the text in the `i`'th
+    column represents a to-do's title that is `i` levels nested in the
+    to-do tree.
+    """
+    with NamedTemporaryFile(mode='w+', delete=False) as file:
+        writer = csv.writer(file)
+        writer.writerow([REMOVE_IF_HEADER, SKIP_IF_HEADER])
+        _write_template_rows(writer, todo_list_template.todos, 0)
+        file.flush()
+        return _upload_csv_to_google(
+            '{} - {}'.format(todo_list_template.name, timezone.now()),
+            file)
+
+
+def import_from_spreadsheet(todo_list_template, spreadsheet_url, request):
+    """Imports a Google Sheet at `spreadsheet_url` to
+    `todo_list_template.todos`.
+
+    For debugging/provenance purposes, creates a
+    TodoListTemplateImportRecord with the user, todo_list_template,
+    and spreadsheet URL behind the import.
+    """
+    reader = get_google_spreadsheet_as_csv(spreadsheet_url, reader=csv.reader)
+    header = next(reader)
+    if header[:2] != [REMOVE_IF_HEADER, SKIP_IF_HEADER]:
+        raise SpreadsheetImportError('Unexpected header: {}'.format(header))
+    # The `i`'th entry in parent_items is current list of child to-dos
+    # of the `i`-depth parent. We use this to determine which parent
+    # to add a child to when we read to-do in a row with lower
+    # indentation than its parent.
+    parent_items = []
+    todos = None
+    for rowindex, row in enumerate(reader):
+        item = {
+            'id': rowindex,
+            'remove_if': json.loads(row[0] or '[]'),
+            'skip_if': json.loads(row[1] or '[]'),
+            'items': []
+        }
+        nonempty_columns = [(columnindex, text)
+                            for columnindex, text in enumerate(row[2:])
+                            if text]
+        if len(nonempty_columns) == 0:
+            continue
+        elif len(nonempty_columns) > 1:
+            raise SpreadsheetImportError(
+                'More than one text entry in row {}: {}'.format(
+                    rowindex, row))
+
+        nonempty_index = nonempty_columns[0][0]
+        item['description'] = nonempty_columns[0][1]
+
+        if todos is None:
+            todos = item
+        elif nonempty_index <= len(parent_items):
+            parent_items = parent_items[:nonempty_index]
+            parent_items[-1].insert(0, item)
+        else:
+            raise SpreadsheetImportError(
+                'Row {} is not a child of a previous row: {}'.format(
+                    rowindex, row))
+        parent_items.append(item['items'])
+
+    todo_list_template.todos = todos
```

Comment:
Overall I find this function pretty tough to read, and in particular it doesn't seem to neatly invert the 'depth' logic you have on export. I wonder if it would be more readable to do this recursively,  like "parse rows of the same depth into a list of items, and recursively build an item's children whenever you encounter a new depth". Maybe you have to iterate over the spreadsheet in reverse to do that?

Some specific questions on the code:
* this line 116 sets `todo_list_template.todos = todos`, but `todos` is just set to the last row in the spreadsheet at this point, right? Is that related to my question above (i.e., does the last row in the spreadsheet correspond to a single root item node for all templates?)
* Lines 107-109 are pretty hard to interpret. At the very least let's add a comment explaining them.
* On line 108, are we losing the parent_items to the right of `nonempty_index`? Or is this the logic that figures out that those items aren't actually parent items, and they're already included in `item`?

Reply:
Thank you for this feedback. I couldn't figure out a nice way to turn my code into a recursive equivalent of the export (I have some vague ideas about using the stack instead of `parent_items`, but it feels more confusing and/or I'm being lazy/intellectually dishonest :P). I took a stab at commenting this close to line-by-line. LMK your thoughts.

Reply:
Okay, let's leave this code. I figured out how to make it recursive, but don't think rewriting it will make it much more readable, so iterative + your great comments is probably good enough.

In case you're curious, my thoughts on a recursive approach:
* The problem is that the tree of items isn't a proper datastructure, it's just a nested python dict. This means that there's no way to pop back 'up' the tree when we encounter a row of depth <= its parent.
* We could solve this by a) implementing a tree datastructure (probably overkill), or b) explicitly passing the parent links into the recursion. In pseudo-code, b roughly looks like:
   ```
   def _recursive_call(all_rows, cur_row_index, parent_stack):
       # all_rows = the csv parsed into a list of rows
       # cur_row_index = the row index we're looking at
       # parent stack = [root_item, root_child_1, root_grandchild_x, ...], all items that are parents of the current row
       cur_row = all_rows[cur_row_index]
       cur_row_depth = find_first_nonempty_row(cur_row)
       cur_item = build_item_dict(cur_row)
       if not parent_stack: # we haven't added the root yet, do so and recurse to the next row
           _recursive_call(all_rows, cur_row_index + 1, [cur_item])
       if cur_row_depth > len(parent_stack):  # we have a child of the latest parent, put it in there and recurse to the next row
           parent_stack[-1]['items'].insert(0, _recursive_call(all_rows, cur_row_index + 1, parent_stack + [cur_item]))
       else: # ignoring the error cases, we are no longer the child of the latest parent, pop up the parent stack and try again with this row
           _recursive_call(all_rows, cur_row_index, parent_stack[:-1])
   
   ```
   * Basically, recursion lets us rewrite your parent_items list to be a more intuitive object, but introduces other complexities that probably balance out the overall pain of the funcion.

Reply:
Awesome! If nothing else, this helps us know it's possible! :)